### PR TITLE
feat: View Action for Speaker

### LINKED ISSUE #4559: View action for speaker not working in dashboard
**Describe the bug**

View action is taking us to edit but it should take us to view like for sessions.

**To Reproduce**
Steps to reproduce the behaviour:
1. Create an event first
2. Then create speaker and session for that event
3. Now go to event overview page (e.g http://localhost:4200/events/e1e58b51)
4. Click on `Speakers` tab
5. Now click on `View` in `Actions` column. See the bug

**Expected behaviour**

It should take us to View page like for sessions.

**Screenshots**

View Speaker Action

**Desktop (please complete the following information):**
 - OS: Linux Mint 20
 - Browser: Firefox

**Additional context**

Will work on this

### PR DESCRIPTION
**Pull request by daretobedifferent18**
Fixes #4559 

#### Short description of what this resolves:
View Action was transiting the page to editing the speaker info, so the issue is resolved.

#### Changes proposed in this pull request:

- Since there is no view page for speakers, so view page is being created.
- The route to which view button was transiting was wrong, so fixed it and now transiting it to view speaker page.
- Improved code for Cell Buttons for speaker.

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia).
- [x] My branch is up-to-date with the Upstream `development` branch.
- [x] The acceptance, integration, unit tests and linter pass locally with my changes 
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

Before
---
Screenshot from 2020-10-27 01-22-02
---
View action button was taking to Edit page.

Screen Shot 2020-10-27 at 01 06 55

After
---
Screenshot from 2020-10-27 01-21-30
---
Now View Action is taking to new view speaker page

Screen Shot 2020-10-27 at 01 08 15
---
on clicking the speaker

Screen Shot 2020-10-27 at 01 08 23

### CODE REVIEW

File: app/controllers/events/view/speakers/list.js [hunk 1]
```diff
@@ -90,8 +90,9 @@ export default class extends Controller.extend(EmberTableControllerMixin) {
   }
 
   @action
-  viewSpeaker(id) {
-    this.transitionToRoute('events.view.speakers.edit', id);
+  viewSpeaker(speaker_id, event) {
+    const event_id = event.router.location.location.pathname.split('/')[2];
```
Comment (iamareebjamal) [inline app/controllers/events/view/speakers/list.js:94]: Use speaker object to get event ID

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 4
           

Complexity increasing per file
==============================
- app/controllers/public/speaker/view.js  3
- app/controllers/events/view/speakers/list.js  1
- tests/unit/routes/public/speaker/view-test.js  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6403732&bid=20757903)

File: app/controllers/public/speaker/view.js [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import Controller from '@ember/controller';
+import { action, computed } from '@ember/object';
```
Comment (niranjan94) [inline app/controllers/public/speaker/view.js:2]: Codacy found an issue: ['action' is defined but never used.](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6403732)

File: app/controllers/public/speaker/view.js [hunk 3]
```diff
@@ -0,0 +1,9 @@
...
+
+export default class extends Controller {
+    @computed('speaker.isAuthenticated', 'authManager.currentUser.email', 'authManager.currentUser.isAdmin', 'model.speakers')
+    get showEditSection() {
+      return this.session.isAuthenticated && (this.authManager.currentUser.isAdmin || this.model.email === this.authManager.currentUser.email);
+    }
```
Comment (niranjan94) [inline app/controllers/public/speaker/view.js:8]: Codacy found an issue: [Expected indentation of 2 spaces but found 4.](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6403732)

File: app/controllers/public/speaker/view.js [hunk 4]
```diff
@@ -0,0 +1,9 @@
+import Controller from '@ember/controller';
+import { action, computed } from '@ember/object';
+
+export default class extends Controller {
+    @computed('speaker.isAuthenticated', 'authManager.currentUser.email', 'authManager.currentUser.isAdmin', 'model.speakers')
+    get showEditSection() {
+      return this.session.isAuthenticated && (this.authManager.currentUser.isAdmin || this.model.email === this.authManager.currentUser.email);
```
Comment (niranjan94) [inline app/controllers/public/speaker/view.js:7]: Codacy found an issue: [Expected indentation of 4 spaces but found 6.](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6403732)

File: app/controllers/public/speaker/view.js [hunk 5]
```diff
@@ -0,0 +1,9 @@
+import Controller from '@ember/controller';
+import { action, computed } from '@ember/object';
+
+export default class extends Controller {
+    @computed('speaker.isAuthenticated', 'authManager.currentUser.email', 'authManager.currentUser.isAdmin', 'model.speakers')
+    get showEditSection() {
```
Comment (niranjan94) [inline app/controllers/public/speaker/view.js:6]: Codacy found an issue: [Expected indentation of 2 spaces but found 4.](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6403732)

File: app/controllers/events/view/speakers/list.js (hunk 1 above)
Comment (daretobedifferent18) [reply app/controllers/events/view/speakers/list.js:94]: Do I have to find record of speaker in this action with the help of speaker_id and extract event id?

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 4
           

Complexity increasing per file
==============================
- app/controllers/public/speaker/view.js  3
- app/controllers/events/view/speakers/list.js  1
- tests/unit/routes/public/speaker/view-test.js  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6403732&bid=20757903)

File: app/controllers/public/speaker/view.js (hunk 2 above)
Comment (niranjan94) [inline app/controllers/public/speaker/view.js:2]: Codacy found an issue: ['action' is defined but never used.](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6403732)

File: app/controllers/public/speaker/view.js (hunk 3 above)
Comment (niranjan94) [inline app/controllers/public/speaker/view.js:8]: Codacy found an issue: [Expected indentation of 2 spaces but found 4.](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6403732)

File: app/controllers/public/speaker/view.js (hunk 4 above)
Comment (niranjan94) [inline app/controllers/public/speaker/view.js:7]: Codacy found an issue: [Expected indentation of 4 spaces but found 6.](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6403732)

File: app/controllers/public/speaker/view.js (hunk 5 above)
Comment (niranjan94) [inline app/controllers/public/speaker/view.js:6]: Codacy found an issue: [Expected indentation of 2 spaces but found 4.](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6403732)

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           

Complexity increasing per file
==============================
- tests/unit/routes/public/speaker/view-test.js  1
- app/controllers/events/view/speakers/list.js  1
- app/controllers/public/speaker/view.js  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6403732&bid=20757903)

File: app/controllers/public/speaker/view.js [hunk 6]
```diff
@@ -0,0 +1,9 @@
+import Controller from '@ember/controller';
+
+export default class extends Controller {
+
+  @computed('speaker.isAuthenticated', 'authManager.currentUser.email', 'authManager.currentUser.isAdmin', 'model.speakers')
```
Comment (niranjan94) [inline app/controllers/public/speaker/view.js:5]: Codacy found an issue: ['computed' is not defined.](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6403732)

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           

Complexity increasing per file
==============================
- app/controllers/events/view/speakers/list.js  1
- app/controllers/public/speaker/view.js  3
- tests/unit/routes/public/speaker/view-test.js  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6403732&bid=20757903)

File: app/controllers/public/speaker/view.js [hunk 7]
```diff
@@ -0,0 +1,10 @@
+import Controller from '@ember/controller';
+import { computed } from '@ember/object';
+
+export default class extends Controller {
+
```
Comment (niranjan94) [inline app/controllers/public/speaker/view.js:5]: Codacy found an issue: [Trailing spaces not allowed.](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6403732)

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- tests/unit/routes/public/speaker/view-test.js  1
- app/controllers/public/speaker/view.js  3
- app/controllers/events/view/speakers/list.js  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6403732&bid=20757903)

**Comment by daretobedifferent18** [comment]
Screenshot from 2020-10-27 17-10-45

File: app/templates/components/ui-table/cell/events/view/speakers/cell-buttons.hbs [hunk 8]
```diff
@@ -1,17 +1,13 @@
 {{this.record}}
-<div class="hidden ui divider"></div>
+<br><br>
```
Comment (iamareebjamal) [inline app/templates/components/ui-table/cell/events/view/speakers/cell-buttons.hbs:2]: Why?

**Comment by daretobedifferent18** [comment]
@iamareebjamal  Actually we can revert this back to 
```html
<div class="hidden ui divider"></div>
``` 
and it will work absolutely fine.
The reason why I did it is that in session file the way of doing the same things is like this so just bcz of similar code I though it would be better but the other way will work fine.

This is for session file
```handlebars
{{this.record}}
<br><br>
<div class="ui horizontal compact basic buttons">
  <UiPopup @content={{t "View Session"}} @class="ui icon button" @click={{action this.props.actions.viewSession this.extraRecords.id this.extraRecords.event.id}} @position="left center">
    <i class="unhide icon"></i>		
  </UiPopup>
  {{#if (not this.extraRecords.isLocked)}}
    <UiPopup @content={{t "Edit Session"}} @class="ui icon button" @click={{action this.props.actions.editSession this.extraRecords.id this.extraRecords.event.id}} @position="left center">
      <i class="edit icon"></i>
    </UiPopup>
  {{/if}}
  <UiPopup @content={{t "Delete Session"}} @click={{action (confirm (t "Are you sure you would like to delete this Session?") (action this.props.actions.deleteSession this.extraRecords.id))}} @class="ui icon button" @position="left center">
    <i class="trash icon"></i>
  </UiPopup>
</div>
```

File: app/controllers/events/view/speakers/list.js [hunk 9]
```diff
@@ -90,9 +90,9 @@ export default class extends Controller.extend(EmberTableControllerMixin) {
   }
 
   @action
-  viewSpeaker(speaker_id, event) {
-    const event_id = event.router.location.location.pathname.split('/')[2];
-    this.transitionToRoute('public.speaker.view', event_id, speaker_id);
+  viewSpeaker(speaker_id) {
```
Comment (iamareebjamal) [inline app/controllers/events/view/speakers/list.js:93]: Why not pass the speaker directly?

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- app/controllers/public/speaker/view.js  3
- app/controllers/events/view/speakers/list.js  1
- tests/unit/routes/public/speaker/view-test.js  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6403732&bid=20757903)

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- app/controllers/public/speaker/view.js  3
- app/controllers/events/view/speakers/list.js  1
- tests/unit/routes/public/speaker/view-test.js  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6403732&bid=20757903)

File: app/controllers/events/view/speakers/list.js (hunk 9 above)
Comment (daretobedifferent18) [reply app/controllers/events/view/speakers/list.js]: I think the way I have done now in latest commit looks fine. 
Please look into it.

File: app/templates/components/ui-table/cell/events/view/speakers/cell-buttons.hbs [hunk 10]
```diff
@@ -1,17 +1,13 @@
...
-    <UiPopup @content={{t "View"}} @class="ui icon button" @click={{action this.props.actions.viewSpeaker this.extraRecords.id}} @position="left center">
-      <i class="unhide icon"></i>
-    </UiPopup>
-  </LinkTo>
-  <LinkTo @route="events.view.speakers.edit" @model={{this.extraRecords.id}}>
-    <UiPopup @content={{t "Edit"}} @class="ui icon button" @click={{action this.props.actions.editSpeaker this.extraRecords.id}} @position="left center">
-      <i class="edit icon"></i>
-    </UiPopup>
-  </LinkTo>
+  <UiPopup @content={{t "View"}} @class="ui icon button" @click={{action this.props.actions.viewSpeaker this.extraRecords.id this.extraRecords.event.id}} @position="left center">
```
Comment (iamareebjamal) [inline app/templates/components/ui-table/cell/events/view/speakers/cell-buttons.hbs:4]: You can just pass this.extraRecords. That's the speaker. And then access speaker.event.id

Why making it so complicated?

Comment (daretobedifferent18) [reply app/templates/components/ui-table/cell/events/view/speakers/cell-buttons.hbs:4]: Yes we can pass just one argument instead of two.

Comment (daretobedifferent18) [reply app/templates/components/ui-table/cell/events/view/speakers/cell-buttons.hbs]: Done

**Review by iamareebjamal — DISMISSED** [review]

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- tests/unit/routes/public/speaker/view-test.js  1
- app/controllers/public/speaker/view.js  3
- app/controllers/events/view/speakers/list.js  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6403732&bid=20757903)

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- tests/unit/routes/public/speaker/view-test.js  1
- app/controllers/public/speaker/view.js  3
- app/controllers/events/view/speakers/list.js  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6403732&bid=20757903)

File: app/controllers/events/view/speakers/list.js [hunk 11]
```diff
@@ -15,7 +15,7 @@ export default class extends Controller.extend(EmberTableControllerMixin) {
       {
         name            : 'Name',
         valuePath       : 'name',
-        extraValuePaths : ['id'],
+        extraValuePaths : ['id', 'event'],
```
Comment (iamareebjamal) [inline app/controllers/events/view/speakers/list.js:18]:
```suggestion
        extraValuePaths : ['id'],
```

File: app/controllers/events/view/speakers/list.js [hunk 12]
```diff
@@ -90,8 +90,8 @@ export default class extends Controller.extend(EmberTableControllerMixin) {
   }
 
   @action
-  viewSpeaker(id) {
-    this.transitionToRoute('events.view.speakers.edit', id);
+  viewSpeaker(speaker) {
+    this.transitionToRoute('public.speaker.view', speaker.event.get('id'), speaker.id);
```
Comment (iamareebjamal) [inline app/controllers/events/view/speakers/list.js:94]:
```suggestion
    this.transitionToRoute('public.speaker.view', speaker.event.get('identifier'), speaker.id);
```

File: app/templates/components/ui-table/cell/events/view/speakers/cell-buttons.hbs [hunk 13]
```diff
@@ -1,17 +1,13 @@
...
+    <i class="unhide icon"></i>		
+  </UiPopup>
+  <UiPopup @content={{t "Edit"}} @class="ui icon button" @click={{action this.props.actions.editSpeaker this.extraRecords.id }} @position="left center">
+    <i class="edit icon"></i>
+  </UiPopup>
   <UiPopup @content={{t "Delete"}} @click={{action (confirm (t "Are you sure you would like to delete this Speaker?") (action this.props.actions.deleteSpeaker this.extraRecords.id))}} @class="ui icon button" @position="left center">
       <i class="trash icon"></i>
   </UiPopup>
-</div>
+</div>
```
Comment (iamareebjamal) [inline app/templates/components/ui-table/cell/events/view/speakers/cell-buttons.hbs:13]:
```suggestion
</div>

```

File: app/templates/public/speaker/view.hbs [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+      {{/if}}
+    </LinkTo>
+  </div>
+  {{/if}}
+  <div class="ui row">
+    <Public::SpeakerItem
+      @speaker={{this.model}}
+    />
+  </div>
+</div>
```
Comment (iamareebjamal) [inline app/templates/public/speaker/view.hbs:20]:
```suggestion
</div>

```

File: app/controllers/events/view/speakers/list.js (hunk 11 above)
Comment (daretobedifferent18) [reply app/controllers/events/view/speakers/list.js]: We need to provide event in it, without this we won't be able to get event id bcz this.extraRecords is not complete speaker object just properties of speaker that are defined in this extraValuePaths.

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- app/controllers/public/speaker/view.js  3
- app/controllers/events/view/speakers/list.js  1
- tests/unit/routes/public/speaker/view-test.js  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-frontend/pullRequest?prid=6403732&bid=20757903)

File: app/controllers/events/view/speakers/list.js (hunk 12 above)
Comment (daretobedifferent18) [reply app/controllers/events/view/speakers/list.js]: Should I make this change? bcz it is working both ways

File: app/controllers/events/view/speakers/list.js [hunk 15]
```diff
@@ -15,7 +15,7 @@ export default class extends Controller.extend(EmberTableControllerMixin) {
       {
         name            : 'Name',
         valuePath       : 'name',
-        extraValuePaths : ['id', 'event'],
+        extraValuePaths : ['id'],
```
Comment (daretobedifferent18) [inline app/controllers/events/view/speakers/list.js]: But this won't work as I said this.extraRecords is not complete speaker object, it has prop of speaker object that are defined in extraValuePaths.

Comment (daretobedifferent18) [reply app/controllers/events/view/speakers/list.js]: @iamareebjamal its not working with this change, please see to it.

**Comment by daretobedifferent18** [comment]
Where will I get to know how many issues I have solved for Fossasia till now?

File: app/controllers/events/view/speakers/list.js (hunk 15 above)
Comment (iamareebjamal) [reply app/controllers/events/view/speakers/list.js]: OK, please send a PR with this change

Comment (daretobedifferent18) [reply app/controllers/events/view/speakers/list.js]: OK and specify same issue #4559 with it?

Comment (iamareebjamal) [reply app/controllers/events/view/speakers/list.js]: Same

Comment (daretobedifferent18) [reply app/controllers/events/view/speakers/list.js]: Please see PR #5405

**Pull request merged**